python_bindings/tutorial/lesson_10_generators_1.py

```diff
@@ -0,0 +1,242 @@
+    input_buffer = hl.InputBuffer(hl.UInt(8), 2)
+    brighten_factor = hl.InputScalar(hl.Float(32))
+
+    # We also declare the Outputs as public member variables, this time of
+    # type hl.OutputScalar or hl.OutputBuffer.
+    #
+    # Each OutputBuffer is filled in as though it were a Func, with the constraint
+    # that it will always be scheduled as .compute_root().
+    brighter = hl.OutputBuffer(hl.UInt(8), 2)
+
+    # We then define a method that constructs and schedules a Halide
+    # pipeline. Each Generator must define this method.
+    def generate(self):
+        # We can declare the Vars we use here, or as class scope, or at global
+        # scope, depending on what is convenient. We'll just declare some here for now.
+        x, y = hl.Var("x"), hl.Var("y")
+
+        # Define the Output(s) as you would for any other Func of Halide code.
+        # We'll mimic the code from lesson 2.
+        self.brighter[x, y] = hl.cast(hl.UInt(8), hl.min(
+            self.input_buffer[x, y] * self.brighten_factor, 255))
+
+        # Also provide whatever schedule you like. As noted above,
+        # all Output(s) will implicitly be calculated as .compute_root(),
+        # whether you explicitly specify that or not.
+        self.brighter.vectorize(x, 16).parallel(y)
+
+
+def use_first_generator():
+    # We can use a Generator with the JIT in several ways; we'll cover just one
+    # of them here. (We'll cover using Generators for ahead-of-time compilation in a later lesson.)
+
+    # The simplest way to use a Generator is to simply invoke its .call()
+    # method; this allows you to "call" a Generator like an ordinary function.
+    #
+    # You provide the Inputs to the Generator as the arguments (specified either
+    # by-name or by-position in the usual Python way; it returns the Outputs
+    # as a Func (or a tuple-of-Funcs if there are multiple Outputs).
+    w, h = 640, 480
+    input_buffer = hl.Buffer(hl.UInt(8), [w, h])
+    input_buffer.fill(128)
+
+    # Using a Generator also requires that there is an active GeneratorContext;
+    # this is an object that contains the Target that the Generator will use,
+    # along with other information we'll learn about in later lessons.
+    # When using a Generator for ahead-of-time compilation, Halide takes
+    # care of setting up a GeneratorContext for you, but when using the JIT,
+    # you'll need to explicitly activate one, using Python's "with" clause:
+    with hl.GeneratorContext(hl.get_jit_target_from_environment()):
+        output_func = MyFirstGenerator.call(input_buffer, brighten_factor=1.5)
+
+    # The result is a Func, so we can just realize() it like any other:
+    output_buffer = output_func.realize([w, h])
+
+    for yy in range(h):
+        for xx in range(w):
+            assert output_buffer[xx, yy] == 192
+
+
+# If you like, you can put multiple Generators in the same source file. This
```

Comment:
Are the first two sentences in this paragraph still helpful in the python context?

Reply:
Nope. Removing.